### PR TITLE
Support passing extra arguments to injected envoy binary

### PR DESCRIPTION
**Pull request by ndhanushkodi**
Changes proposed in this PR:
- Based on work from #177 
- Supports passing flags to the envoy binary in 2 ways
    1. as a flag through the consul-k8s inject-connect command
    2. as a pod annotation

You can set it via either one, but if you set the args both ways, the pod annotation arguments will be used.

How I've tested this PR:
- Built a docker image (`gcr.io/nitya-293720/consul-k8s-dev:extraEnvoyArgs`) and ran it on a cluster, while passing `--log-level debug` as an extra envoy flag, and saw that when we deployed a connect-inject enabled service, the envoy logs were showing debug logs.
- Set the pod annotation to be a different set of arguments, and saw that the arguments passed to the envoy binary were overwritten.

How I expect reviewers to test this PR:
- Code review

Checklist:
- [x] Tests added
- [x] CHANGELOG entry added (*HashiCorp engineers only, community PRs should not add a changelog entry*)

### CODE REVIEW

**Review by ashwin-venkatesh — COMMENTED** [review]
This looks great. Have a few minor questions but love this overall!!

File: subcommand/inject-connect/command.go [hunk 1]
```diff
@@ -109,6 +110,8 @@ func (c *Command) init() {
 		"Docker image for Envoy. Defaults to envoyproxy/envoy-alpine:v1.13.0.")
 	c.flagSet.StringVar(&c.flagConsulK8sImage, "consul-k8s-image", "",
 		"Docker image for consul-k8s. Used for the connect sidecar.")
+	c.flagSet.StringVar(&c.flagEnvoyExtraArgs, "envoy-extra-args", "",
+		"Extra envoy command line args to be set when starting envoy (e.g \"--log-level debug\").")
```
Comment (lkysow) [inline subcommand/inject-connect/command.go:114]: drive-by: should say how to pass multiple args

**Review by ndhanushkodi — COMMENTED** [review]
Re-requesting review: addressed PR comments, and added functionality to pass the envoy binary arguments via pod annotations as well.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,13 @@
 ## UNRELEASED
 
+IMPROVEMENTS:
+* Connect: support passing extra arguments to the envoy binary. [[GH-378](https://github.com/hashicorp/consul-k8s/pull/378)]
+    Arguments can be passed in 2 ways:
+    * via a flag to the consul-k8s inject-connect command:
+      e.g. `consul-k8s inject-connect -envoy-extra-args="--log-level debug --disable-hot-restart"`
+    * via pod annotations
+      e.g. `consul.hashicorp.com/envoy-extra-args:"--log-level debug --disable-hot-restart"`
```
Comment (ishustava) [inline CHANGELOG.md:9]:
```suggestion
      e.g. `consul.hashicorp.com/envoy-extra-args: "--log-level debug --disable-hot-restart"`
```

File: connect-inject/handler.go [hunk 3]
```diff
@@ -85,6 +85,13 @@ const (
 	annotationSidecarProxyMemoryLimit   = "consul.hashicorp.com/sidecar-proxy-memory-limit"
 	annotationSidecarProxyMemoryRequest = "consul.hashicorp.com/sidecar-proxy-memory-request"
 
+	// annotationEnvoyExtraArgs is a space-separated list of arguments to be passed to the
+	// envoy binary. See list of args here: https://www.envoyproxy.io/docs/envoy/latest/operations/cli
+	// e.g. consul.hashicorp.com/envoy-extra-args: "--log-level debug --disable-hot-restart"
+	// The arguments passed in via this annotation will take precendence over arguments
+	// passed via the helm template value `connectInject.envoyExtraArgs`
```
Comment (ishustava) [inline connect-inject/handler.go:92]:
```suggestion
	// passed via the -envoy-extra-args flag.`
```

**Review by ishustava — APPROVED** [review]
Looks great 🎉 💃 ! I've left a couple of edits, but they are non-blocking.

File: connect-inject/envoy_sidecar.go [hunk 4]
```diff
@@ -87,6 +91,38 @@ func (h *Handler) envoySidecar(pod *corev1.Pod, k8sNamespace string) (corev1.Con
...
+		"--config-path", "/consul/connect-inject/envoy-bootstrap.yaml",
+	}
+
+	extraArgs, annotationSet := pod.Annotations[annotationEnvoyExtraArgs]
+
+	if annotationSet || h.EnvoyExtraArgs != "" {
+
+		extraArgsToUse := h.EnvoyExtraArgs
+
+		// prefer args set by pod annotation over the flag to the consul-k8s binary (h.EnvoyExtraArgs)
```
Comment (ashwin-venkatesh) [inline connect-inject/envoy_sidecar.go:106]:
```suggestion
		// Prefer args set by pod annotation over the flag to the consul-k8s binary (h.EnvoyExtraArgs).
```

File: connect-inject/envoy_sidecar.go [hunk 5]
```diff
@@ -87,6 +91,38 @@ func (h *Handler) envoySidecar(pod *corev1.Pod, k8sNamespace string) (corev1.Con
...
+		if annotationSet {
+			extraArgsToUse = extraArgs
+		}
+
+		// split string into tokens
```
Comment (ashwin-venkatesh) [inline connect-inject/envoy_sidecar.go:111]:
```suggestion
		// Split string into tokens.
```

**Review by ashwin-venkatesh — APPROVED** [review]
🥳 Brilliant!!
